### PR TITLE
CORDA-1117 Fix non deterministic port allocation issue in ExplorerSimulation

### PR DESCRIPTION
**Pull request by kchalkias**
Port allocation in `ExplorerSimulation` is using Driver and previous implementation was not deterministic regarding port allocation.
A fix by implementing a custom [PortAllocation] has been applied.
Docs and readme files have been updated with the correct fixed ports.

### CODE REVIEW

File: testing/node-driver/src/main/kotlin/net/corda/testing/driver/Driver.kt [hunk 1]
```diff
@@ -107,18 +106,24 @@ data class WebserverHandle(
...
-    /** Get the next available port via [nextPort] and then return a [NetworkHostAndPort] **/
+    /** Get the next available port via [nextPort] and then return a [NetworkHostAndPort]. */
     fun nextHostAndPort() = NetworkHostAndPort("localhost", nextPort())
 
     /**
-     * An implementation of [PortAllocation] which allocates ports sequentially
+     * This method is required from custom [PortAllocation] implementations in which port allocation policy
+     * is based on the [input] object. Default implementation calls [nextHostAndPort].
+     */
+    open fun <T : Any> nextHostAndPort(input: T) = nextHostAndPort()
```
Comment (shamsasari) [inline testing/node-driver/src/main/kotlin/net/corda/testing/driver/Driver.kt]: We shouldn't be exposing a public API method just to fix a demo.

File: testing/node-driver/src/main/kotlin/net/corda/testing/node/internal/DriverDSLImpl.kt [hunk 2]
```diff
@@ -184,7 +184,7 @@ class DriverDSLImpl(
             startInSameProcess: Boolean?,
             maximumHeapSize: String
     ): CordaFuture<NodeHandle> {
-        val p2pAddress = portAllocation.nextHostAndPort()
+        val p2pAddress = portAllocation.nextHostAndPort(Pair(providedName, "p2p"))
```
Comment (shamsasari) [inline testing/node-driver/src/main/kotlin/net/corda/testing/node/internal/DriverDSLImpl.kt]: This new API isn't designed right. The type of the argument is `Any` but the implementation here is clearly using `Pair<CordaX500Name, String>` and so you anyone wanting to provide their own implementation would need to know that.

File: testing/node-driver/src/main/kotlin/net/corda/testing/node/internal/DriverDSLImpl.kt [hunk 3]
```diff
@@ -215,10 +215,10 @@ class DriverDSLImpl(
                                     customOverrides: Map<String, Any?>,
                                     startInSameProcess: Boolean? = null,
                                     maximumHeapSize: String = "200m",
-                                    p2pAddress: NetworkHostAndPort = portAllocation.nextHostAndPort()): CordaFuture<NodeHandle> {
-        val rpcAddress = portAllocation.nextHostAndPort()
-        val rpcAdminAddress = portAllocation.nextHostAndPort()
-        val webAddress = portAllocation.nextHostAndPort()
+                                    p2pAddress: NetworkHostAndPort = portAllocation.nextHostAndPort(Pair(name, "p2p"))): CordaFuture<NodeHandle> {
+        val rpcAddress = portAllocation.nextHostAndPort(Pair(name, "rpc"))
```
Comment (shamsasari) [inline testing/node-driver/src/main/kotlin/net/corda/testing/node/internal/DriverDSLImpl.kt]: These strings should be enums

**Review by shamsasari — CHANGES_REQUESTED** [review]
This is the wrong approach. I still think we can use `CordformDefinition` to define upfront the ports. Have a look at `BankOfCordaCordformTest` which runs the BOC demo as a CordformDefinition by calling `deployAndRunNodesThen` which allows you to run your own code after all the nodes have started.

**Comment by kchalkias** [comment]
I had a solution with enums, wanted to agree first on the api. Although I feel the flexibility for custom `PortAllocation` will be useful in the future anyway, will change implementation to use `CordformDefinition`

**Pull request closed without merging**